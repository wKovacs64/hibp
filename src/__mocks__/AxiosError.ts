/** @internal */
export default class AxiosError extends Error {
  constructor(response) {
    super();
    this.response = response;
  }
}
