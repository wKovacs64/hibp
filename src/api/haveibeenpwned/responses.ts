/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v3#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

import type { ResponseBody } from './types';

/** @internal */
export interface HaveIBeenPwnedApiResponse {
  headers: Map<string, string>;
  status: number;
  statusText?: string;
  body?: ResponseBody;
}

const emptyHeaders = new Map<string, string>();

/** @internal */
export const OK: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 200,
};

/** @internal */
export const BAD_REQUEST: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 400,
  statusText:
    'Bad request â€” the account does not comply with an acceptable format.',
};

/**
 * This response has unique behavior. For some reason, the API includes an
 * object in the response body for this one, containing a human-readable
 * message. Manually populating the message here purely for use in tests.
 *
 * @internal
 */
export const UNAUTHORIZED: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 401,
  body: {
    statusCode: 401,
    message: 'Access denied due to missing hibp-api-key.',
  },
};

/** @internal */
export const FORBIDDEN: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 403,
  statusText: 'Forbidden - access denied.',
};

/** @internal */
export const BLOCKED: HaveIBeenPwnedApiResponse = {
  headers: new Map([['cf-ray', 'someRayId']]),
  status: 403,
};

/** @internal */
export const NOT_FOUND: HaveIBeenPwnedApiResponse = {
  headers: emptyHeaders,
  status: 404,
};

/**
 * This response has unique behavior. For some reason, the API includes an
 * object in the response body for this one, containing a human-readable
 * message. Manually populating the message here purely for use in tests.
 *
 * @internal
 */
export const TOO_MANY_REQUESTS: HaveIBeenPwnedApiResponse = {
  headers: new Map([['retry-after', '2']]),
  status: 429,
  body: {
    statusCode: 429,
    message: 'Rate limit is exceeded. Try again in 2 seconds.',
  },
};
