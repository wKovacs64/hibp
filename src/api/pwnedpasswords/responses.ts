/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/api/v3#PwnedPasswords
 *
 */

/** @internal */
export interface PwnedPasswordsApiResponse {
  status: number;
  body?: string;
}

/** @internal */
export const OK: PwnedPasswordsApiResponse = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST: PwnedPasswordsApiResponse = {
  status: 400,
  body: 'The hash prefix was not in a valid format',
};
