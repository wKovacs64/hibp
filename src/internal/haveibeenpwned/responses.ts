/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v2#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

/** @internal */
export const OK = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST = {
  status: 400,
  statusText:
    'Bad request â€” the account does not comply with an acceptable format.',
};

/** @internal */
export const FORBIDDEN = {
  status: 403,
  statusText: 'Forbidden - no user agent has been specified in the request.',
};

/** @internal */
export const NOT_FOUND = {
  status: 404,
};

/**
 * This response has unique behavior. For some reason, the API includes a
 * human-readable message in the response body for this one. Manually populating
 * the message here purely for use in tests.
 *
 * @internal
 */
export const TOO_MANY_REQUESTS = {
  status: 429,
  data:
    'Rate limit exceeded, refer to acceptable use of the API: ' +
    'https://haveibeenpwned.com/API/v2#AcceptableUse',
};
