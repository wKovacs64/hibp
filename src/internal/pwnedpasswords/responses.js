/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/API/v2#PwnedPasswords
 *
 */

export const RANGE_OK = {
  status: 200,
};

export const RANGE_BAD_REQUEST = {
  status: 400,
  response: 'The hash prefix was not in a valid format',
};
