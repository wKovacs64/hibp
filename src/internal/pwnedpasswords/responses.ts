/**
 * Known potential responses from the remote API.
 *
 * https://haveibeenpwned.com/API/v2#PwnedPasswords
 *
 */

export interface PwnedPasswordsApiResponse {
  // eslint-disable-next-line no-restricted-globals
  status: number;
  data?: string;
}

/** @internal */
export const OK = {
  status: 200,
};

/** @internal */
export const BAD_REQUEST = {
  status: 400,
  data: 'The hash prefix was not in a valid format',
};
