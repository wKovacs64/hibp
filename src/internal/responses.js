/**
 * Known potential responses from the remote API.
 *
 * Unfortunately, the API does not send a decent human-readable message back
 * with each response, but they are documented on the website:
 * https://haveibeenpwned.com/api/v2#ResponseCodes
 *
 * These objects simply provide a mapping between the HTTP response status code
 * and the corresponding human-readable message so we can throw a more
 * descriptive error for the consumer. (They are also leveraged in our tests.)
 */

export const OK = {
  status: 200,
};

export const BAD_REQUEST = {
  status: 400,
  statusText: 'Bad request â€” the account does not comply with an acceptable ' +
    'format.',
};

export const FORBIDDEN = {
  status: 403,
  statusText: 'Forbidden - no user agent has been specified in the request.',
};

export const NOT_FOUND = {
  status: 404,
  statusText: 'Not found - the account could not be found and has therefore ' +
    'not been pwned.',
};

/**
 * This response has unique behavior. For some reason, the API includes a
 * human-readable message in the response body for this one. Manually populating
 * the message here purely for use in tests.
 */
export const TOO_MANY_REQUESTS = {
  status: 429,
  response: 'Rate limit exceeded, refer to acceptable use of the API: ' +
    'https://haveibeenpwned.com/API/v2#AcceptableUse',
};
