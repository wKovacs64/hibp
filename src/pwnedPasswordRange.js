import fetchFromApi from './internal/pwnedpasswords/fetchFromApi';

/**
 * Fetches the SHA-1 hash suffixes for the given 5-character SHA-1 hash prefix.
 *
 * When a password hash with the same first 5 characters is found in the Pwned
 * Passwords repository, the API will respond with an HTTP 200 and include the
 * suffix of every hash beginning with the specified prefix, followed by a count
 * of how many times it appears in the data set.
 * This function parses the response and returns a more structured format.
 *
 * @param {string} prefix the first 5 characters of a SHA-1 password hash (case
 * insensitive)
 * @returns {Promise} a Promise which resolves to an array of objects each
 * containing the `suffix` that when matched with the prefix composes the
 * complete hash, and a `count` of how many times it appears in the breached
 * password data set, or rejects with an Error
 *
 * @example
 * const suffix = '1E4C9B93F3F0682250B6CF8331B7EE68FD8';
 * pwnedPasswordRange('5BAA6')
 *   // filter to matching suffix
 *   .then(
 *     arr => arr.filter(item => item.suffix === suffix)
 *   )
 *   // return count if match, 0 if not
 *   .then(
 *     arr => (arr[0] ? arr[0].count : 0)
 *   )
 *   .catch((err) => {
 *     // ...
 *   });
 * @see https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange
 * @alias module:pwnedPasswordRange
 */
const pwnedPasswordRange = prefix =>
  fetchFromApi(`/range/${encodeURIComponent(prefix)}`)
    // each line to an array
    .then(res => res.split('\n'))
    // each line split into suffix and count
    .then(arr =>
      arr.map(item => ({
        suffix: item.split(':')[0],
        count: parseInt(item.split(':')[1], 10),
      })),
    );

/**
 * A module for determining if a password's SHA-1 hash has been exposed in a
 * breach.
 *
 * @module pwnedPasswordRange
 * @example
 * import { pwnedPasswordRange } from 'hibp';
 */
export default pwnedPasswordRange;
