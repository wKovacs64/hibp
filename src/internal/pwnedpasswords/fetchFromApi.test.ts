import AxiosError from 'AxiosError';
import pwnedPasswordRange from 'pwnedPasswordRange';
import { BAD_REQUEST } from './responses';
import mockAxios from './axiosInstance';

describe('internal (pwnedpassword): fetchFromApi', () => {
  describe('request failure', () => {
    it('re-throws request setup errors', () => {
      const ERR = new Error('Set sail for fail!');
      mockAxios.get.mockRejectedValueOnce(ERR);
      expect(pwnedPasswordRange('setup error')).rejects.toEqual(ERR);
    });
  });

  describe('invalid range', () => {
    it('throws a "Bad Request" error', () => {
      mockAxios.get.mockRejectedValueOnce(new AxiosError(BAD_REQUEST));
      expect(pwnedPasswordRange('bad request')).rejects.toMatchSnapshot();
    });
  });

  describe('unexpected HTTP error', () => {
    it('throws an error with the response status text', () => {
      mockAxios.get.mockRejectedValueOnce(
        new AxiosError({
          status: 999,
          statusText: 'Unknown - something unexpected happened.',
        }),
      );
      expect(pwnedPasswordRange('unknown response')).rejects.toMatchSnapshot();
    });
  });
});
