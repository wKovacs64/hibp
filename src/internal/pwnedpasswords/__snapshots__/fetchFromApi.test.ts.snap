// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internal (pwnedpassword): fetchFromApi invalid range throws a "Bad Request" error 1`] = `[Error: The hash prefix was not in a valid format]`;

exports[`internal (pwnedpassword): fetchFromApi unexpected HTTP error throws an error with the response status text 1`] = `[Error: Unknown - something unexpected happened.]`;
